### ISSUE TITLE
training on voxceleb1+2 is very slow?

### ISSUE THREAD

**Issue by dragen1860** [speechbrain/speechbrain#990]
Dear all:
  I noticed that when training on voxceleb1+2, it will take me up to 25 hours for single epoch. and even with ddp on 4 gpu cards, the training speed does not reduce at all. I guess the cpu is the bottleneck? anyone has the same phenomena? thank you.

**Comment by mravanelli** [comment]
It should be much faster (see the logs that we share with google drive in
the readme). Are you copying the data in the local compute node or you are
reading from the shared filesystem?

**Comment by dragen1860** [comment]
@mravanelli Thanks for your reply. I read from local SSD. It's very fast because i used it to train ImageNet. But when train on voxceleb1+2 1M wav file, it is very slow.

**Comment by dragen1860** [comment]
@mravanelli by the way. Where is the google drive link, I cannot find it. Thank you.

**Comment by mravanelli** [comment]
It should be in the readme file

**Comment by dragen1860** [comment]
sorry, during to certain reasons, I can not .

**Comment by pnsafari** [comment]
Hi @dragen1860, did you find the google drive link which was mentioned by @mravanelli?

**Comment by mravanelli** [comment]
The location of all the model+logs is now reported in the README file. You
can find it here:
https://drive.google.com/file/d/1EziERcHD_gyE6qc8DbxPKU1isVf7pbNl/view?usp=sharing

**Comment by pnsafari** [comment]
Thank you @mravanelli. Regarding the parallel training, I tried with 4 and 6 GPUs (on a single machine). They are Tesla V100. The expected training time is about 4h in both cases. I use different numbers of workers for instance 4, 8, 48 and it does not change that much. The strange behavior that I cannot understand is for the first 12% of the training it is done in only a few seconds and then it becomes slow and the expected time would be about 4h. Just to mention I try to train only over vox2. Just note that I increase the batch size to 256 for the parallel case. Any idea @TParcollet @ZiRaymondal ? I didn't modify anything in the architecture.

**Comment by mravanelli** [comment]
Titouan found the issue why this recipe is slower when using ddp. He will
push the fix soon (it is related to cudnn benchmark=false if I recall
correctly).

**Comment by TParcollet** [comment]
@pnsafari Are you using DP or DDP ?

**Comment by pnsafari** [comment]
@TParcollet Just DP.

**Comment by TParcollet** [comment]
So what I found should not impact what you are experiencing, it's only impacting DDP. Note however that with DP, you MUST increase the batch size accordingly to your number of GPUs as it is dividing the batch_size by nb_gpus. Otherwise, no speed_up. DP also doesn't scale well above 3-4 GPUs in my experience. Let me know if you wanna try with DDP, I'm working on the fix right now (just to comment two lines ahah).

**Comment by pnsafari** [comment]
Yes, I increased to 256 according to the number of GPUs. Maybe I ll also switch to DDP. I didn't try DDP since I didn't expect to get much since I am working on a single node.

**Comment by TParcollet** [comment]
DDP is superior to DP even for single node training, however, see my latest PR about this, you will have to remove two lines of code from the core.py file.

**Comment by mravanelli** [comment]
@pnsafari, please let us know if the PR by @TParcollet solves the issue. We are conducting some tests as well.

**Comment by pnsafari** [comment]
@mravanelli, I have switched to DDP as @TParcollet proposed. Now it is working with batch size 32  on vox2 dataset, and with 4 workers the `ecapa-tdnn` is showing an expected training time of less than 3h/epoch. I ll update you if I face some error or strange behavior. BTW, this is the command I used:
`CUDA_VISIBLE_DEVICES=1,2,3,4,5,6 python -m torch.distributed.launch --nproc_per_node=6 train_speaker_embeddings.py hparams/train_ecapa_tdnn.yaml --distributed_launch --distributed_backend nccl`

**Comment by TParcollet** [comment]
@pnsafari could you try to set the cudnn.benchmark to true in core.py (instead of removing it) ? We are investigating this right now :-) It could speed even further the training

**Comment by pnsafari** [comment]
@TParcollet I would like to do so, but I am struggling with releasing the GPUs. The job was killed but still I got the error that GPUs are not available. I am looking for some zombie processes but I cannot find them. I ll try with those modifications as soon as I can release the GPUs :)

**Comment by pnsafari** [comment]
@TParcollet I tried all combinations for `~ 60 70` iterations. It seems the only case which makes everything slow is while `cudnn.deterministic = True`. The other options wouldn't change anything that much. I think it is fine to remove `cudnn.deterministic` and just keep the `cudnn.benchmark = True`.

**Comment by TParcollet** [comment]
@mravanelli interesting. @pnsafari thanks!

**Comment by lamnguyenx** [comment]
I've tried limiting the maximum threads spawned by torch and achieved x6 times faster training on A100 GPU:
```
import torch
torch.set_num_threads(16)
```

Seems that it was due to too many threads spawning on high-end CPU (we are using AMD EPYC 7742)

**Comment by Adel-Moumen** [comment]
Hello,

It seems that the issue has been answered. Therefore, I am closing this issue. If you have any other questions, please feel free to reopen the issue! Thanks. 🙂
